doc/devices/drivers/pinctrl/pinctrl.rst

```diff
@@ -0,0 +1,1186 @@
+..
+    Copyright (c) 2018 Bobby Noelte
+    SPDX-License-Identifier: Apache-2.0
+
+.. _device_drivers_pinctrl:
+
+PINCTRL (pin controller) device driver
+######################################
+
+.. contents::
+   :depth: 1
+   :local:
+   :backlinks: top
+
+The **pin controller** device driver:
+
+- enumerates controllable **pins**,
+- multiplexes **pins**, and
+- configures **pins**.
+
+The Zephyr *pin controller* follows the concepts of the Linux PINCTRL
+(PIN CONTROL) subsystem linux_pinctrl_. The following documentation
+provides a summary of the concepts and definitions available in detail
+in the Linux documentation ( linux_pinctrl_, linux_pinctrl_bindings_ ).
+The Zephyr *pin controller* does not provide the full scope of the Linux
+PINCTRL subsystem. Also the PINCTRL interface is specific to Zephyr and
+not compatible to the Linux PINCTRL interface. However the device tree
+bindings are kept compatible to the linux_pinctrl_bindings_ as much as
+possible.
+
+Pin controller concepts
+***********************
+
+A **PINCTRL device driver**
+    provides access to the *pin controller*.
+
+A **pin controller**
+    is a set of hardware resources, usually accessed via registers, that
+    can control *pins*. It may be able to multiplex, bias, set load capacitance,
+    set drive strength, etc. for individual *pins* or *groups* of pins.
+    There may be several pin controllers in a system which control only a set of
+    the available *pins*.
+
+A **pin**
+    is the physical input/ output line of a chip. Pins are denoted by u16_t
+    integer numbers. The pin numberspace is local [#pin_number_space]_ to
+    a *pin controller*. This pin space may be sparse - i.e. there may be gaps
+    in the space with numbers where no pin exists.
+
+A pins **group**
+    is a set of *pins* that are can be multiplexed to a
+    common *function*. Groups are denoted by u16_t integer numbers. The group
+    numberspace is local to each *pin controller*. The group number space may
+    be sparse or even non-existing.
+
+A pinctrl **state**
+    defines the *pin configuration* and the *pin multiplexing* to a *function*
+    of a *group* of *pins*. Pinctrl states are denoted by u16_t integer numbers.
+    The pinctrl state number space is local to a *pin controller*.
+
+A **function**
+    is a logical or physical function block of a chip. Functions are
```

Comment:
To ease reading, can you provide an example of physical functional block at this point?